core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java

```diff
@@ -0,0 +1,41 @@
+package io.substrait.extendedexpression;
+
+import io.substrait.expression.Expression;
+import io.substrait.proto.AdvancedExtension;
+import io.substrait.proto.AggregateFunction;
+import io.substrait.type.NamedStruct;
+import java.util.List;
+import java.util.Optional;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class ExtendedExpression {
+  public abstract List<ExpressionReference> getReferredExpressions();
+
+  public abstract NamedStruct getBaseSchema();
+
+  public abstract List<String> getExpectedTypeUrls();
+
+  // creating simple extensions, such as extensionURIs and extensions, is performed on the fly
+
+  public abstract Optional<AdvancedExtension> getAdvancedExtension();
+
+  @Value.Immutable
+  public abstract static class ExpressionReference {
+    public abstract ExpressionTypeReference getExpressionType();
```

Comment:
I think that having the ExpressionReference contain the type of the reference within itself is a bit clunky as a design. Looking at the `ExtendedExpressionProtoConverter`, we have to do instanceof checks to figure out the type and then make decision based on the type.

I think a cleaner way to approach this is with inheritance with something like:
```java
  public interface ExpressionReferenceBase {
    public List<String> getOutputNames();

    public io.substrait.proto.ExpressionReference toProto(...);
  }

  @Value.Immutable
  public abstract static class ExpressionReference implements ExpressionReferenceBase {
    public abstract Expression getExpression();

    @Override
    public io.substrait.proto.ExpressionReference toProto(...) {
      return ...
    }
  }

  @Value.Immutable
  public abstract static class AggregateFunctionReference implements ExpressionReferenceBase {
    public abstract Aggregate.Measure getMeasure();

    @Override
    public io.substrait.proto.ExpressionReference toProto(ExpressionProtoConverter converter) {
      return null;
    }
  }
```
effectively, each class knows how to convert itself.

@davisusanibar what do you think?